taipy/core/reason/reasons.py

```diff
@@ -0,0 +1,62 @@
+# Copyright 2021-2024 Avaiga Private Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+# the License. You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+# an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+# specific language governing permissions and limitations under the License.
+
+from typing import Dict, Set
+
+from .reason import Reason
+
+
+class Reasons:
+    """
+    This class is used to store all the reasons to explain why some Taipy operations are not allowed.
+
+    Because Taipy applications are natively multiuser, asynchronous, and dynamic,
+    some functions might not be called in some specific contexts. You can protect
+    such calls by calling other methods that return a Reasons object. It acts like a
+    boolean: True if the operation can be performed and False otherwise.
+    If the action cannot be performed, the Reasons object holds all the `reasons as a list
+    of `Reason` objects. Each `Reason` holds an explanation of why the operation cannot be
+    performed.
+    """
+
+    def __init__(self, entity_id: str) -> None:
```

Comment:
probably to know which entity that reason belongs to?

Reply:
An action like `submit` may have multiple failures, on different entities, so I think the `entity_id` here is "reason"able

Reply:
A ReasonCollection object is not necessarily attached to an entity but to a boolean function call. Some individual Reason objects can be attached to an entity, though.

For example, `can_create(Scenario_cfg)` is not attached to any entity.
We can also imagine adding reasons to any function/method unrelated to an entity, like `is_standalone_mode_on()`, `is_core_running()`, etc. 

Because this ReasonCollection object is not persisted, keeping the entity_id is unnecessary, IMHO. Plus, I checked, and it is not read anywhere else but in the tests.

Am I missing something? What do you think?

Reply:
no you're right, let me remove that then